### PR TITLE
Manage listing categories in dashboard

### PR DESCRIPTION
**Pull request by Rafi993**
## What type of PR is this? 

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

It adds ability to add/update/remove listing categories in admin panel

## Related Tickets & Documents

closes https://github.com/forem/forem/issues/10481

## QA Instructions, Screenshots, Recordings

https://vimeo.com/463735841/dfe6705dab

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.forem.com
- [ ] readme
- [x] no documentation needed

### CODE REVIEW

File: app/lib/constants/role.rb [hunk 1]
```diff
@@ -18,6 +18,7 @@ module Role
                      "Resource Admin: Config",
                      "Resource Admin: Broadcast",
                      "Resource Admin: HtmlVariant",
-                     "Resource Admin: DisplayAd"].freeze
+                     "Resource Admin: DisplayAd",
+                     "Resource Admin: ListingCategory"].freeze
```
Comment (Rafi993) [inline app/lib/constants/role.rb:22]: I am not sure if it should be called Listing or ListingCategory

Comment (rhymes) [reply app/lib/constants/role.rb:22]: as it's a category, I think `ListingCategory` is correct

**Comment by Rafi993** [comment]
I think some unrelated test is failing

**Comment by rhymes** [comment]
@Rafi993 restarted, cross fingers :D

File: app/controllers/admin/listing_categories_controller.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    end
+
+    def destroy
+      @listing_category = ListingCategory.find(params[:id])
+
+      if @listing_category.destroy
+        flash[:success] = "Listing Category has been deleted!"
+        redirect_to admin_listing_categories_path
+      else
+        flash[:danger] = "Something went wrong with deleting the Listing Category."
```
Comment (fdoxyz) [inline app/controllers/admin/listing_categories_controller.rb:54]: It's best to do the same as the `update` action so the error message will be descriptive

```suggestion
        flash[:danger] = @listing_category.errors.full_messages.to_sentence
```

**Review by fdoxyz — CHANGES_REQUESTED** [review]
Thank you for the PR @Rafi993, great job in following the sample PR from the issue! 

Functionality-wise the PR is working great locally.

I have a small change I think it would be good to include and another non-blocking suggestion. I checked the sample PR from the issue and just now noticed this one slipped in there, so not your fault at all!

With this tiny change in place I'd say this looks good.

**Comment by Rafi993** [comment]
> Thank you for the PR @Rafi993, great job in following the sample PR from the issue! 
> 
> Functionality-wise the PR is working great locally.
> 
> I have a small change I think it would be good to include and another non-blocking suggestion. I checked the sample PR from the issue and just now noticed this one slipped in there, so not your fault at all!
> 
> With this tiny change in place I'd say this looks good.

Thank you @fdoxyz I'll make those changes.

**Review by fdoxyz — APPROVED** [review]
Awesome thanks again @Rafi993!

The codeclimate errors can be ignored since they're complaining about similar code in different places. So far we've considered it's best to maintain a similar approach in different admin controllers and maybe refactor later on in a way that would make this go away (or get rid of the codeclimate rule for this particular scenario).

**Review by juliannatetreault — COMMENTED** [review]
Thank you so much for adding Listing Categories management to `/admin`, @Rafi993! I tested everything out locally and it all worked really well. One thing that I did notice though, is that the placeholder text in the search bar, found under `/admin/listings/categories`, on a 15-inch screen on Google Chrome is _slightly_ cut off (shown in the picture below). That aside, everything looked and worked great! 🎉 
Screen Shot 2020-10-01 at 12 54 38 PM

**Review by mstruve — APPROVED** [review]
Tested, works great!

**Comment by Rafi993** [comment]
> Thank you so much for adding Listing Categories management to `/admin`, @Rafi993! I tested everything out locally and it all worked really well. One thing that I did notice though, is that the placeholder text in the search bar, found under `/admin/listings/categories`, on a 15-inch screen on Google Chrome is _slightly_ cut off (shown in the picture below). That aside, everything looked and worked great! 
> Screen Shot 2020-10-01 at 12 54 38 PM

Thank you @juliannatetreault I did try to emulate 15 inch mac with and without retina display in chrome dev tools with following values

* size 2880 x 1800
* I tried device pixel ration 1, 2 and 3 (default value for my screen is 2)
* My screen size is 14 inches hdpi screen

 I also tried couple of other screen resolutions too but was unable to reproduce this issue.

**Comment by fdoxyz** [comment]
I do get the same effect on 13 inch chrome, but also ran into this over with DisplayAds

Screen Shot 2020-10-01 at 20 39 54

I couldn't find a way to make the input grow horizontally there (I tried a few things from https://storybook.forem.com/).

A suggestion: Shortening the placeholder to "Search by name" would make it fit (like in DisplayAds) and it wouldn't be cut out.
Maybe nice to have: A fix for the element to grow horizontally (could also be used to enhance other dashboards like this one)

**Comment by Rafi993** [comment]
> I do get the same effect on 13 inch chrome, but also ran into this over with DisplayAds
> 
> Screen Shot 2020-10-01 at 20 39 54
> 
> I couldn't find a way to make the input grow horizontally there (I tried a few things from https://storybook.forem.com/).
> 
> A suggestion: Shortening the placeholder to "Search by name" would make it fit (like in DisplayAds) and it wouldn't be cut out.
> Maybe nice to have: A fix for the element to grow horizontally (could also be used to enhance other dashboards like this one)

Thank you @fdoxyz I have changed the text to "Search by name" now.

**Comment by Rafi993** [comment]
Thank you @fdoxyz  should I revert the "search by name" placeholder change?

**Comment by fdoxyz** [comment]
So sorry @Rafi993 that's my bad! I was checking in to see if CI had failed on mobile and closed by accident 🙈

I'd love to give the opportunity for other different timezones to weigh in, we may get a suggestion for manipulating the input field a little better. It still has my approval 🙌

**Comment by ludwiczakpawel** [comment]
There's not an easy way (right now!) to manipulate easily the width of the input.. Changing the label to be shorter is a way to go IMO: "Search by name" and call it a day :)

We can of course do some hacky things but I don't think it's worth it :)

**Review by juliannatetreault — APPROVED** [review]
Thank you for making the change to the placeholder text, @Rafi993! I think this PR is good-to-go now. Thanks again for doing this! 🎉

**Comment by Rafi993** [comment]
> Thank you for making the change to the placeholder text, @Rafi993! I think this PR is good-to-go now. Thanks again for doing this!

Thank you @juliannatetreault

**Pull request merged**